File: typescript/src/sort.test.ts
import { expect } from "chai";
import { sort as bubbleSort } from "./bubble";
import { sort as heapSort } from "./heap";
import { sort as mergeSort } from "./merge";
import { sort as quicksort } from "./quicksort";
import { sort as selectionSort } from "./selection";

describe("Sort", () => {
  [
    {name: "Selection", sort: selectionSort},
    {name: "Bubble", sort: bubbleSort},
    {name: "Quicksort", sort: quicksort},
    {name: "Merge", sort: mergeSort},
    {name: "Heap", sort: heapSort},
  ].forEach(({name, sort}) => {
    context(`${name}`, () => {
      context("for empty array", () => {
        it("returns an empty array", (done) => {
          expect(sort([])).to.have.lengthOf(0);
          done();
        });
      });

      context("for sorted array", () => {
        let sortedArray: number[];
        before((done) => {
          sortedArray = [1, 2, 3, 4, 5, 6];
          done();
        });

        it("returns same array", (done) => {
          expect(sort(sortedArray)).to.have.lengthOf(sortedArray.length);
          expect(sort(sortedArray)).to.eql(sortedArray);
          done();
        });
      });

      context("for not sorted array", () => {
        [
          {
            sorted: [1, 2, 3, 4, 5, 11, 88, 99],
            unsorted: [3, 5, 2, 4, 1, 99, 88, 11],
          },
          {
            sorted: [1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6],
            unsorted: [1, 2, 3, 4, 5, 6, 1, 2, 3, 4, 5, 6],
          },
        ].forEach(({unsorted, sorted}) => {
          it(`[${unsorted.join(", ")}] returns sorted array`, (done) => {
            expect(sort(unsorted)).to.have.lengthOf(unsorted.length);
            expect(sort(unsorted)).to.eql(sorted);
            done();
          });
        });
      });
    });
  });
});

File: kotlin/src/test/kotlin/in/kanur/sorts/SortTest.kt
package `in`.kanur.sorts

import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import kotlin.test.assertTrue

object SortTest: Spek({
  arrayOf(Selection, Bubble, Quicksort, Merge, HeapSort).forEach { implementation ->
    describe("${implementation::class.simpleName}.sort") {
      context("when input is empty") {
        val testArray: IntArray by memoized { intArrayOf() }

        it("returns empty array") {
          assertTrue {
            testArray.contentEquals(implementation.sort(testArray))
          }
        }
      }

      context("when input is sorted") {
        val testArray: IntArray by memoized { intArrayOf(1,2,3) }

        it("returns the same array") {
          assertTrue {
            testArray.contentEquals(implementation.sort(testArray))
          }
        }
      }

      context("when input is array of same number") {
        val testArray: IntArray by memoized { intArrayOf(1,1,1) }

        it("returns the same array") {
          assertTrue {
            testArray.contentEquals(implementation.sort(testArray))
          }
        }
      }

      context("when input is not sorted") {
        arrayOf(
          Pair(
            intArrayOf(3,2,1),
            intArrayOf(1,2,3)
          ),
          Pair(
            intArrayOf(3, 5, 2, 4, 1, 99, 88, 11),
            intArrayOf(1, 2, 3, 4, 5, 11, 88, 99)
          ),
          Pair(
            intArrayOf(1, 2, 3, 4, 5, 6, 1, 2, 3, 4, 5, 6),
            intArrayOf(1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6)
          )
        ).forEach { (unsorted, sorted) ->
          it("returns the sorted array") {
            assertTrue {
              sorted.contentEquals(implementation.sort(unsorted))
            }
          }
        }
      }
    }
  }
})
